fairscale/experimental/nn/ssd_offload.py

```diff
@@ -68,6 +69,20 @@ def read(input_tensor: torch.Tensor, filename: str, file_offset_bytes: int = 0)
             assert data_read == chunk_end - chunk_start
 
 
+class StorageState(Enum):
+    """
+    Simple enum to indicate whether the tensor handle is pointing
+    to data on disk or memory. This is useful for asserting on 
+    whether the tensor is available for operations or if it needs
+    to be moved from disk to CPU or device.
+    """
+
+    UNALLOCATED = auto()
+    ON_DISK = auto()
+    ON_CPU = auto()
+    ON_GPU = auto()
+
+
 class SsdTensorHandle(torch.Tensor):
     """
     This class extends from torch.Tensor and represents a Tensor which is backed by SSD storage.
```

Comment:
super late to the party, but could that relate to the "meta" tensor discussions (virtual device) which are being developed in current pytorch ? worth being in touch if it's not already the case ? I see that as a specialized meta tensor basically

Reply:
We've implemented this in collaboration/discussion with the PyTorch team. This was the recommended way for us to implement a new Tensor type where we could essentially hook into the dispatch process. We prototyped a meta tensor implementation as part of adding support that had its own issues.
The meta tensor you mention is also related for sure but for our purposes is required more from an initialization POV. We will be using the new [`init_meta`](https://github.com/pytorch/pytorch/pull/66317) API for our delayed initialization needs.